src/playground/sample_data_ingestion/sample_data_plugins/sample_data_plugin.py

```diff
@@ -0,0 +1,145 @@
+from datetime import datetime, timedelta
+from os import environ, path, remove
+from json import dumps, load, loads
+from shutil import copyfileobj
+from requests import request
+from base64 import b64encode
+from time import mktime
+import gzip
+
+
+"""
+Generic Plugin class: initialization is defined here
+Arguments:
+    - index_name: The name of the source index
+    - payload: The JSON body of which to call create API requests
+    - base_url: The IP address/url where the OpenSearch backend lies
+"""
+class Plugin:
+    def __init__(self, index_name, payload, base_url):
+        self.index_name = index_name
+        self.id = ""
+        self.payload = payload
+        
+        # Convert to JSON string if necessary
+        if type(self.payload) is dict:
+            self.payload = dumps(self.payload)
+        
+        # Add authentication headers
+        __auth_plain = environ.get("SAMPLE_DATA_USERNAME") + ":" + environ.get("SAMPLE_DATA_PASSWORD")
+        __auth = b64encode(__auth_plain.encode("ascii"))
+        __authorization = "Basic " + str(__auth).split("\'")[1]
+        self.headers = {
+            'Authorization': __authorization,
+            'Content-Type': 'application/json'
+        }
+
+        # Build url
+        self.base_url = base_url
+        if self.base_url.endswith("/"):
+          self.base_url = self.base_url[:-1]
+        self.url = self.base_url + ":9200/"
+    
+    # If payload is a zipped config file, unzip it and return unzipped filename
+    def unzip(self):
+        if type(self.payload) is not str or (type(self.payload) is str and ".json" not in self.payload):
+            print("Payload is not a filename or is not a .json file")
+            return None
+        if type(self.payload) is str and ".gz" not in self.payload:
+            print("Payload is already unzipped")
+            return self.payload
+        with gzip.open(self.payload, 'rt') as fin:
+            with open(self.payload.split(".gz")[0], 'wb') as fout:
+                copyfileobj(fin, fout)
+        return self.payload.split(".gz")[0]
+
+    # Utility function to convert payload to a dict
+    def convert_payload(self, payload, plugin_config_directory_name):
+        if type(payload) is str and ".json" not in payload:
```

Comment:
does it mean that payload is a file but not a json file??

Reply:
> +1, you could have a base file for plugin class, and then have separate files for different plugins

Done

> does it mean that payload is a file but not a json file??

Comments are updated but this code block sees that this payload is a string (non filename) and will try to convert to a `dict` (because the function assumes the `payload` is a `JSON string`). If this fails, the function knows the input was a non `JSON string`, non `filename string` string.